README.md
# WordPress Admin Style
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fbueltge%2Fwordpress-admin-style.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2Fbueltge%2Fwordpress-admin-style?ref=badge_shield)
[![Latest Stable Version](https://poser.pugx.org/bueltge/wordpress-admin-style/v/stable)](https://packagist.org/packages/bueltge/wordpress-admin-style)

Shows the WordPress admin styles on one page to help you to develop WordPress compliant

 * List markup and examples for '2 Column Page Layout', Headers, Header Icons, Buttons, Tables, Admin Notices, 
Alternative Colours, Pagination, Form Elements, Form Helper Functions, Tabs, Default Admin Color Schemes
 * [WP Color Picker](https://make.wordpress.org/core/2012/11/30/new-color-picker-in-wp-3-5/)
 * Helper Classes
 * jQuery UI Examples
 * [Dashicons](https://developer.wordpress.org/resource/dashicons/)
 * [Genericons](http://genericons.com/)

## Description
WordPress is developing fast - this also applies to the design of the backend. 
So it is important not to use your own styles in the admin area and use tags and classes of WordPress. 
This is the best way you can simplify your work as a developer, and you don't have to test the design with every update. 
Unfortunately, there are quite extensive opportunities in the backend to implement the requirements. 
Several different classes and HTML structures are used. To be able to look up something this simple, 
I have developed this small plugin, which tinkers in the development environment and quickly represents the necessary elements. 

Also see the [post](https://wpengineer.com/2226/new-plugin-to-style-your-plugin-on-wordpress-admin-with-default-styles/) 
about the idea and also the comments for a discussion.

### Screenshots & history

![Part of the Admin Style information in WordPress version 5.7](./assets/screenshot-8.png)

The plugin is still active in usage and development since an early WordPress version, so you should see it on the screenshots from the different versions below.

 1. [Admin Style in WordPress 3.1](/assets/screenshot-1.png)
 2. [Admin Style in WordPress 3.4](/assets/screenshot-2.png)
 3. [Admin Style in WordPress 3.6-alpha](/assets/screenshot-3.png)
 4. [Admin Style in WordPress MP6 Design](/assets/screenshot-4.png)
 5. [Dashicons in WordPress MP6 Design](/assets/screenshot-5.png)
 6. [Genericons in WordPress 3.9-alpha](/assets/screenshot-6.png)
 7. [Admin Style in WordPress 5.3](/assets/screenshot-7.png)

## Installing
### Manual
 1. Download the plugin as a [ZIP file](https://github.com/bueltge/WordPress-Admin-Style/archive/master.zip) from GitHub.
 2. In your WordPress admin click *Plugin -> Add New -> Upload Plugin*.
 3. Upload the ZIP file.
 4. Activate the plugin.

### Via Composer
The plugin is available as [Composer package](https://packagist.org/packages/bueltge/wordpress-admin-style) and can be installed via Composer from the root of your WordPress installation.

`composer require bueltge/wordpress-admin-style`

If you have changed the default directory structure or just want the plugin to a specific location, you can create a project from the Composer package.

`composer create-project bueltge/wordpress-admin-style <optional-name>`

### Via Git
 1. Go to your Must Use folder `cd path`
 2. `git init .`
 3. `git remote add origin https://github.com/bueltge/wordpress-admin-style.git`
 4. `git fetch origin`
 5. `git checkout master`
 6. Check in the network plugin page, if it works

### Automatic Udates
The plugin supports the [GitHub Updater plugin](https://github.com/afragen/github-updater) for WordPress. The plugin 
enables automatic updates from this GitHub Repository. You will find all information about the how and why at the [plugin wiki page](https://github.com/afragen/github-updater/wiki).

## Other Notes
### Bugs, technical hints or contribute
Please give me feedback, contribute and file technical bugs on this 
[GitHub Repo](https://github.com/bueltge/WordPress-Admin-Style), use Issues.

### License
Good news, this plugin is free for everyone! Since it's released under the [GPLv2+](./LICENSE), 
You can use it free of charge on your personal or commercial blog. But if you enjoy this plugin, 
you can thank me and leave a 
[small donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6069955 "Paypal Donate link") 
for the time I've spent writing and supporting this plugin. 
And I really don't want to know how many hours of my life this plugin has already eaten ;)

### Contact & Feedback
The plugin is designed and developed by me ([Frank Bültge](https://bueltge.de))

Please let me know if you like the plugin, or you hate it or whatever ... 
Please fork it, add an issue for ideas and bugs.

### Disclaimer
I'm German, and my English might be gruesome here and there. 
So please be patient with me and let me know of typos or grammatical farts. Thanks


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

## [1.6.1](https://github.com/bueltge/wordpress-admin-style/compare/1.6.0...1.6.1) - 2023-12-22
* Add a link to the new function wp_admin_notice() documentation (available since 6.4) - Fixes Update Admin Notices #67
* Add a columns to display notices with the css class .notice-alt examples
* Fix demo page Javascript execution (in footer & defer) - Fix jQuery features not working #64
* Fix some HTML errors on scroll to top buttons (probably caused by phpcsfix on commit 92c86f8)
* Fix jQuery UI image sources (unable to open files)
* Reactivate demo page custom CSS

## [1.6.0](https://github.com/bueltge/wordpress-admin-style/compare/1.5.6...1.6.0) - 2021-05-28
* Maintenance scripts to run with jQuery 3.5*
* Maintenance of Prism library to the last stable version.
* Enhance a note for the Datepicker Style from John, #32.
* Add color palette support.
* Add WP Color Picker example, #40

## [1.5.6](https://github.com/bueltge/wordpress-admin-style/compare/1.5.5...1.5.6) - 2020-07-17
* Add new Dashicons', #61

## [1.5.5](https://github.com/bueltge/wordpress-admin-style/compare/1.5.4...1.5.5) - 2020-07-17
* Add new element 'Card', #60
* Leave hints for the Dashicon project.

## [1.5.4](https://github.com/bueltge/wordpress-admin-style/compare/1.5.3...1.5.4) - 2019-07-19
* Update color overview
* Add color table for 'Auxiliary Hues', probs @stefanpejcic

## [1.5.3](https://github.com/bueltge/wordpress-admin-style/compare/1.5.2...1.5.3) - 2019-07-17
* Add ressource for the Gutenberg project, the new editor

## 1.5.2 - 2018-06-15
* Fixed markup in 2 Columns Layout: static (px) section since WP change to button markup instead of div.
* Fixed the admin notices to leave always visible.

## 1.5.1
* Fixed overlay on menu. #46

## [1.5.0](https://github.com/bueltge/wordpress-admin-style/compare/1.4.2...1.5.0) - 2017-05-24
* Added accessibility improvements on h1, in core since WP 4.8, [see make post](https://make.wordpress.org/core/2017/05/17/cleaner-headings-in-the-admin-screens/)
* Small Cleanup source formatting, class name PSR rule.
* Change to an simpler updater for the github repo, Kudos [GitHub Updater (Lite)](https://github.com/FacetWP/github-updater-lite)

## [1.4.2](https://github.com/bueltge/wordpress-admin-style/compare/1.4.1...1.4.2) - 2016-07-06
### Added
* Clarification of how and where the admin notices get moved
* `.inline` example for admin notices
* added `code` tags to admin notices output
 
## [1.4.1](https://github.com/bueltge/wordpress-admin-style/compare/1.4.0...1.4.1) - 2016-07-06
### Fixed
*  English + minor formatting

## [1.4.0](https://github.com/bueltge/wordpress-admin-style/compare/1.3.3...1.4.0) - 2016-05-20
### Added
* Add syntax-highlighting to each pattern.


WpAdminStyle.php
<?php /** @noinspection HtmlUnknownTarget */
/**
 * Plugin Name:       WordPress Admin Style
 * Plugin URI:        https://github.com/bueltge/wordpress-admin-style
 * Text Domain:       WpAdminStyle
 * Domain Path:       /languages
 * GitHub URI:        https://github.com/bueltge/wordpress-admin-style
 * GitHub Plugin URI: https://github.com/bueltge/wordpress-admin-style
 * Description:       Shows the WordPress admin styles on one page to help you to develop WordPress compliant.
 * Author:            Frank Bültge
 * Version:           1.6.1
 * Licence:           GPLv2+
 * License URI:       ./LICENSE
 * Author URI:        https://bueltge.de
 * Last Change:       2023-12-22
 */

! defined( 'ABSPATH' ) && exit;

add_action(
	'plugins_loaded',
	array( WpAdminStyle::get_instance(), 'plugin_setup' )
);

/**
 * Class WpAdminStyle
 */
class WpAdminStyle {

	/**
	 * Characters there we replace in the files.
	 *
	 * @var array
	 */
	protected static $file_replace = array( '.php', '_', '-', ' ' );
	/**
	 * Directory of patters of examples.
	 *
	 * @var string
	 */
	protected $patterns_dir = '';

	/**
	 * Constructor
	 *
	 * @since  0.0.1
	 */
	public function __construct() {
	}

	/**
	 * Points the class, singleton.
	 *
	 * @access public
	 * @since  0.0.1
	 */
	public static function get_instance() {
		static $instance;

		if ( null === $instance ) {
			$instance = new self();
		}

		return $instance;
	}

	/**
	 * Used for regular plugin work.
	 *
	 * @wp-hook  plugins_loaded
	 * @since    05/02/2013
	 */
	public function plugin_setup() {
		$this->load_classes();

		if ( ! is_admin() ) {
			return null;
		}

		$this->patterns_dir = plugin_dir_path( __FILE__ ) . 'patterns';

		// add menu item incl. the example source.
		add_action( 'admin_menu', array( $this, 'add_menu_page' ) );
		// Plugin love.
		add_filter( 'plugin_row_meta', array( $this, 'donate_link' ), 10, 2 );
	}

	/**
	 * Scans the plugins subfolder and include files.
	 *
	 * @return  void
	 * @since   05/02/2013
	 */
	protected function load_classes() {

		// Load required classes.
		foreach ( glob( __DIR__ . '/inc/*.php' ) as $path ) {
			/** @noinspection PhpIncludeInspection */
			require_once $path;
		}
	}

	/**
	 * Add Menu item on WP Backend
	 *
	 * @return void
	 * @since  0.0.1
	 * @uses   add_menu_page
	 * @access public
	 */
	public function add_menu_page() {
		$page_hook_suffix = add_menu_page(
			esc_html__( 'WordPress Admin Style', 'WpAdminStyle' ),
			esc_html__( 'Admin Style', 'WpAdminStyle' ),
			'read',
			'WordPress_Admin_Style',
			array( $this, 'get_style_examples' )
		);
		add_action( 'admin_print_scripts-' . $page_hook_suffix, array( $this, 'enqueue_assets' ) );
	}

	/**
	 * Echo Markup examples
	 *
	 * @return void
	 * @since  0.0.1
	 * @uses
	 * @access public
	 */
	public function get_style_examples() {
		?>

		<div class="wrap">

			<h1><?php echo esc_html( $this->get_plugin_data( 'Name' ) ); ?></h1>

			<?php
			$this->get_mini_menu();
			$files = $this->get_patterns();

			// Load files and get data for view and list source.
			foreach ( $files as $file ) {
				$anker    = str_replace( self::$file_replace, '', $file );
				$patterns = $this->patterns_dir . '/' . $file;

				echo '<section class="pattern" id="' . esc_attr( $anker ) . '">';
				/** @noinspection PhpIncludeInspection */
				include_once $patterns;
				echo '<details class="primer" style="display: inline-block; width: 100%;">';
				echo '<summary title="Show markup and usage">&#8226;&#8226;&#8226; '
					. esc_attr__( 'Show markup and usage', 'WpAdminStyle' )
					. '</summary>';
				echo '<section>';
				echo '<pre><code class="language-php">';
				echo htmlspecialchars(
					file_get_contents(
						$patterns,
						FILE_USE_INCLUDE_PATH
					),
					ENT_QUOTES
				);
				echo '</code></pre>';
				echo '</section>';
				echo '</details><!--/.primer-->';
				echo '<p>';
				echo '<a class="alignright button" href="javascript:void(0);" onclick="window.scrollTo(0,0);" style="margin:3px 0 0 30px;">' . esc_attr__(
					'scroll to top',
					'WpAdminStyle'
				)
					. '</a><br class="clear" />';
				echo '</p>';
				echo '</section><!--/.pattern-->';
				echo '<hr>';
			}
			?>

		</div> <!-- .wrap -->
		<?php
	}

	/**
	 * Return plugin comment data.
	 *
	 * @param string $value default = 'Version'
	 *  also possible is: Name, PluginURI, Version, Description, Author,
	 *                    AuthorURI, TextDomain, DomainPath, Network, Title.
	 *
	 * @return string
	 * @uses   get_plugin_data
	 * @access public
	 * @since  0.0.1
	 */
	private function get_plugin_data( $value = 'Version' ) {
		static $plugin_data = array();

		// fetch the data just once.
		if ( isset( $plugin_data[ $value ] ) ) {
			return $plugin_data[ $value ];
		}

		if ( ! function_exists( 'get_plugin_data' ) ) {
			require_once ABSPATH . '/wp-admin/includes/plugin.php';
		}

		$plugin_data = get_plugin_data( __FILE__ );

		return empty( $plugin_data[ $value ] ) ? '' : $plugin_data[ $value ];
	}

	/**
	 * Print the mini menu for a short navigation.
	 */
	public function get_mini_menu() {
		$patterns = $this->get_patterns( 'headers' );
		?>
		<div id="poststuff">
			<div id="post-body" class="metabox-holder columns-2">

				<!-- main content -->
				<div id="post-body-content">

					<div class="meta-box-sortables ui-sortable">

						<div class="postbox">

							<h2><span><?php esc_attr_e( 'MiniMenu', 'WpAdminStyle' ); ?></span></h2>

							<div class="inside">

								<table class="widefat">
									<?php
									$class = '';
									foreach ( $patterns as $pattern ) {
										$class = '' === $class ? $class = ' class="alternate"' : '';
										$anker = str_replace( self::$file_replace, '', $pattern );

										?>
										<tr<?php echo $class; ?>>
											<td class="row-title">
												<a href="#<?php echo esc_attr( $anker ); ?>">
													<?php echo esc_attr( ucfirst( $pattern ) ); ?>
												</a>
											</td>
										</tr>
										<?php
									} // end foreach patterns
									?>
								</table>

							</div>
							<!-- .inside -->

						</div>
						<!-- .postbox -->

					</div>
					<!-- .meta-box-sortables .ui-sortable -->

				</div>
				<!-- post-body-content -->

				<!-- sidebar -->
				<div id="postbox-container-1" class="postbox-container">

					<div class="meta-box-sortables">

						<div class="postbox">

							<h2><span><?php esc_attr_e( 'About the plugin', 'WpAdminStyle' ); ?></span></h2>

							<div class="inside">
								<p>
									<?php
									printf(
									// translators: The two strings will be replaced by the url to the development repository and the post about the idea.
										__(
											'Please read more about this small plugin on <a href="%1$s">github</a> or in <a href="%2$s">this post</a> on the blog of WP Engineer.',
											'WpAdminStyle'
										),
										'https://github.com/bueltge/WordPress-Admin-Style',
										'http://wpengineer.com/2226/new-plugin-to-style-your-plugin-on-wordpress-admin-with-default-styles/'
									);
									?>
								</p>

								<p>&copy; Copyright 2008 - <?php echo esc_attr( date( 'Y' ) ); ?>
									<a href="https://bueltge.de">Frank Bültge</a></p>
							</div>

						</div>
						<!-- .postbox -->

						<div class="postbox">

							<h2><span><?php esc_attr_e( 'Resources & Reference', 'WpAdminStyle' ); ?></span></h2>

							<div class="inside">
								<ul>
									<li>
										<a href="https://developer.wordpress.org/block-editor/designers/">Editor
											'Gutenberg' Designer Documentation, include patterns and resources</a>
									</li>
									<li>
										<a href="http://dotorgstyleguide.wordpress.com/">WordPress.org UI Style
											Guide</a>
									</li>
									<li>
										<a href="https://make.wordpress.org/core/handbook/best-practices/coding-standards/html/">HTML
											Coding Standards</a>
									</li>
									<li>
										<a href="https://make.wordpress.org/core/handbook/best-practices/coding-standards/css/">CSS
											Coding Standards</a>
									</li>
									<li>
										<a href="https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/">PHP
											Coding Standards</a>
									</li>
									<li>
										<a href="https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/">JavaScript
											Coding Standards</a>
									</li>
									<li><a href="https://make.wordpress.org/design/">WordPress UI Group</a></li>
								</ul>
							</div>

						</div>
						<!-- .postbox -->

					</div>
					<!-- .meta-box-sortables -->

				</div>
				<!-- #postbox-container-1 .postbox-container -->

			</div>
			<br class="clear">
		</div>
		<?php
	}

	/**
	 * Return list of pattern files or name of files
	 *
	 * @param string $type Type of patters, default '', possible is 'headers'.
	 *
	 * @param bool   $sort
	 *
	 * @return array|mixed
	 * @since 2015-03-25
	 */
	public function get_patterns( $type = '', $sort = true ) {
		$files              = array();
		$this->patterns_dir = plugin_dir_path( __FILE__ ) . 'patterns';
		$handle             = opendir( $this->patterns_dir );

		while ( false !== ( $file = readdir( $handle ) ) ) {
			if ( false !== stripos( $file, '.php' ) ) {
				$files[] = $file;
			}
		}

		if ( $sort ) {
			sort( $files );
		}

		$files_h = str_replace( self::$file_replace, ' ', $files );

		if ( 'headers' === $type ) {
			return $files_h;
		}

		return $files;
	}

	/**
	 * Add donate link to plugin description in /wp-admin/plugins.php
	 *
	 * @param array  $plugin_meta All met data to a plugin.
	 * @param string $plugin_file The main file of the plugin with the meta data.
	 *
	 * @return array
	 */
	public function donate_link( $plugin_meta, $plugin_file ) {
		if ( plugin_basename( __FILE__ ) === $plugin_file ) {
			$plugin_meta[] = sprintf(
				'&hearts; <a href="%s">%s</a>',
				'https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6069955',
				esc_html__( 'Donate', 'WpAdminStyle' )
			);
		}

		return $plugin_meta;
	}

	/**
	 * Register and enqueue script and styles for highlighting source.
	 *
	 * @since  2016-05-20
	 */
	public function enqueue_assets() {
		wp_register_style(
			'prism',
			plugins_url( 'css/prism.css', __FILE__ ),
			array(),
			'2021-05-28',
			'screen'
		);
		wp_enqueue_style( 'prism' );

		wp_register_script(
			'prism',
			plugins_url( 'js/prism.js', __FILE__ ),
			array(),
			'2021-05-28',
			true
		);
		wp_enqueue_script( 'prism' );

		wp_register_script(
			'example-color-picker',
			plugins_url( 'js/color-picker.js', __FILE__ ),
			array( 'jquery', 'wp-color-picker' ),
			'2021-05-28',
			true
		);
		wp_enqueue_script( 'example-color-picker' );
	}
} // end class
